File: src/core/pipeline/send_report/handler.py
import io
import os
from datetime import datetime, timezone

import openpyxl
import resend

resend.api_key = os.environ["RESEND_API_KEY"]
EMAIL_SENDER = os.environ["EMAIL_SENDER"]
EMAIL_RECIPIENT = os.environ["EMAIL_RECIPIENT"]

# Column headers and their corresponding product dict keys
COLUMNS = [
    ("Nombre", "name"),
    ("Marca", "brand"),
    ("Precio", "price"),
    ("Precio Original", "original_price"),
    ("URL", "url"),
    ("Fuente", "source"),
    ("Categoría", "category"),
    ("Volumen (ml)", "volume_ml"),
    ("ABV (%)", "abv"),
    ("Fecha Scraping", "scraped_at"),
]


def handler(event, context):
    """
    Generate an Excel report of unmatched products and send it via email.

    Input (batch mode — output of SyncWithApiParallel Map):
        event["results"]: list of sync_with_api result dicts, one per batch
    Input (legacy):
        event["unmatched"]: list of ScrapedProduct dicts

    Returns:
        { "sent": bool, "count": int }
    """
    results: list[dict] = event.get("results", [])
    if results:
        unmatched: list[dict] = []
        for r in results:
            unmatched.extend(r.get("unmatched", []))
    else:
        unmatched = event.get("unmatched", [])

    if not unmatched:
        print("[SendReport] No unmatched products — skipping email")
        return {"sent": False, "count": 0}

    try:
        excel_bytes = _build_excel(unmatched)
        _send_email(excel_bytes, len(unmatched))
        print(f"[SendReport] Sent report with {len(unmatched)} unmatched product(s)")
        return {"sent": True, "count": len(unmatched)}

    except Exception as e:
        print(f"[SendReport] Error: {e}")
        return {"sent": False, "count": len(unmatched), "error": str(e)}


def _build_excel(products: list[dict]) -> bytes:
    wb = openpyxl.Workbook()
    ws = wb.active
    ws.title = "Sin match"

    ws.append([col[0] for col in COLUMNS])

    for product in products:
        ws.append([product.get(key, "") for _, key in COLUMNS])

    buffer = io.BytesIO()
    wb.save(buffer)
    return buffer.getvalue()


def _send_email(excel_bytes: bytes, count: int):
    date_str = datetime.now(timezone.utc).strftime("%Y-%m-%d")

    params: resend.Emails.SendParams = {
        "from": EMAIL_SENDER,
        "to": [EMAIL_RECIPIENT],
        "subject": f"[RDC Scraper] {count} producto(s) sin match — {date_str}",
        "html": f"""
        <h2>Productos sin match — {date_str}</h2>
        <p>Se encontraron <strong>{count} producto(s)</strong> que no pudieron ser
        asociados a ningún item de la base de datos de bebidas.</p>
        <p>Adjunto el Excel con el detalle para revisión y carga manual.</p>
        <hr/>
        <p style="color:#888; font-size:12px;">Generado automáticamente por RDC Scraper</p>
        """,
        "attachments": [
            {
                "filename": f"unmatched_{date_str}.xlsx",
                "content": list(excel_bytes),
            }
        ],
    }

    resend.Emails.send(params)

File: src/pipeline/send-report/handler.ts
import ExcelJS from 'exceljs'
import { Resend } from 'resend'
import type { ScrapedProduct } from '@rdc/spider'

const resend = new Resend(process.env['RESEND_API_KEY'])

export const handler = async (event: {
  results?: Array<{ unmatched?: ScrapedProduct[] }>
  unmatched?: ScrapedProduct[]
}) => {
  const unmatched: ScrapedProduct[] = event.results
    ? event.results.flatMap(r => r.unmatched ?? [])
    : (event.unmatched ?? [])

  if (unmatched.length === 0) return { sent: false, count: 0 }

  const workbook = new ExcelJS.Workbook()
  const sheet = workbook.addWorksheet('Unmatched Products')

  sheet.columns = [
    { header: 'Name',          key: 'name',      width: 40 },
    { header: 'Brand',         key: 'brand',     width: 20 },
    { header: 'Price',         key: 'price',     width: 12 },
    { header: 'Original Price',key: 'bestPrice', width: 14 },
    { header: 'URL',           key: 'url',       width: 60 },
    { header: 'Source',        key: 'source',    width: 12 },
    { header: 'Category',      key: 'category',  width: 14 },
    { header: 'Volume (ml)',   key: 'volumeMl',  width: 12 },
    { header: 'ABV (%)',       key: 'abv',       width: 10 },
    { header: 'Scraped Date',  key: 'date',      width: 14 },
  ]

  const today = new Date().toISOString().split('T')[0]
  for (const p of unmatched) {
    sheet.addRow({ ...p, date: today })
  }

  const buffer = await workbook.xlsx.writeBuffer()

  await resend.emails.send({
    from: process.env['EMAIL_SENDER']!,
    to:   process.env['EMAIL_RECIPIENT']!,
    subject: `RDC Scraper — ${unmatched.length} productos sin match`,
    html: `<p>${unmatched.length} productos no pudieron ser asociados a una bebida.</p>`,
    attachments: [{
      filename: 'unmatched.xlsx',
      content: Buffer.from(buffer).toString('base64'),
    }],
  })

  return { sent: true, count: unmatched.length }
}
